control/prometheus.py

```diff
@@ -0,0 +1,284 @@
+        start_http_server(port)
+    REGISTRY.register(NVMeOFCollector(spdk_rpc_client, config))
+
+
+class NVMeOFCollector:
+    """Provide a prometheus endpoint for nvmeof gateway statistics"""
+
+    def __init__(self, spdk_rpc_client, config):
+        self.logger = logging.getLogger(__name__)
+        self.spdk_rpc_client = spdk_rpc_client
+        self.metric_prefix = "ceph_nvmeof"
+        self.gw_config = config
+        _bdev_pools = config.get_with_default('gateway', 'prometheus_bdev_pools', '')
+        self.bdev_pools = _bdev_pools.split(',') if _bdev_pools else []
+
+        self.start = time.time()
+        self.interval = config.getint_with_default('gateway', 'prometheus_stats_inteval', 10)
+        self.lock = threading.Lock()
+        self.event = threading.Event()
+        self.spdk_version = {}
+        self.bdev_info = []
+        self.bdev_io_stats = {}
+        self.spdk_thread_stats = {}
+        self.subsystems = []
+        self.method_timings = {}
+
+        if self.bdev_pools:
+            self.logger.info(f"Stats restricted to bdevs in the following pool(s): {','.join(self.bdev_pools)}")
+        else:
+            self.logger.info("Stats for all bdevs will be provided")
+
+        self.collector = threading.Thread(target=self._collect)
+        self.collector.daemon = True
+        self.collector.start()
+
+    @timer
+    def _get_version(self):
+        self.spdk_version = rpc.spdk_get_version(self.spdk_rpc_client)
+
+    @timer
+    def _get_bdev_info(self):
+        self.bdev_info = rpc.bdev.bdev_get_bdevs(self.spdk_rpc_client)
+
+    @timer
+    def _get_bdev_io_stats(self):
+        self.bdev_io_stats = rpc.bdev.bdev_get_iostat(self.spdk_rpc_client)
+
+    @timer
+    def _get_spdk_thread_stats(self):
+        self.spdk_thread_stats = rpc.app.thread_get_stats(self.spdk_rpc_client)
+        
+    @timer
+    def _get_subsystems(self):
+        self.subsystems = rpc.nvmf.nvmf_get_subsystems(self.spdk_rpc_client)
+
+    def _collect(self):
+        """Collector thread refreshing the SPDK stats"""
+        self.logger.info(f"Starting SPDK collector thread, refreshing every {self.interval} secs")
+
+        while not self.event.wait(self.interval):
+            with self.lock:
+                self.logger.debug("Refreshing stats from SPDK")
+                start_time = time.time()
+                self._get_version()
+                self._get_bdev_info()
+                self._get_bdev_io_stats()
+                self._get_spdk_thread_stats()
+                self._get_subsystems
+                elapsed = time.time() - start_time
+
+                interval_used = elapsed / self.interval
+                if interval_used > 1:
+                    self.logger.error(f"Stats refresh time > interval time of {self.interval} secs")
+                elif interval_used > 0.8:
+                    self.logger.warning("Stats refresh is close to exceeding the interval (>80%)")
```

Comment:
Not a big deal, but I wouldn't hard-code literals within the core:

```suggestion
                if elapsed > self.interval:
                    self.logger.error(f"Stats refresh time > interval time of {self.interval} secs")
                elif elapsed > COLLECT_ELAPSED_WARNING * self.interval:
                    self.logger.warning(f"Stats refresh is close to exceeding the interval (>{COLLECT_ELAPSED_WARNING:.0%}%)")
```

Reply:
Happy to add this - thanks